app/controllers/LongerCommodityCodeController.scala

```diff
@@ -52,79 +51,74 @@ class LongerCommodityCodeController @Inject() (
-        case _                                                                                                       =>
-          form
-      }
-
-      originalComcodeOpt match {
-        case Some(shortCommodity) if commodityCodeSansTrailingZeros(shortCommodity).length == 6 =>
-          Ok(view(preparedForm, mode, commodityCodeSansTrailingZeros(shortCommodity), recordId))
-        case _                                                                                  =>
+      val previousAnswerOpt = request.userAnswers.get(LongerCommodityCodePage(recordId))
+      val shortComcodeOpt   = getShortCommodityCodeOpt(recordId, request, validLength)
+      val preparedForm      = previousAnswerOpt.map(answer => form.fill(answer)).getOrElse(form)
+
+      shortComcodeOpt match {
+        case Some(shortComcode) if shortComcode.length == validLength =>
+          Ok(view(preparedForm, mode, shortComcode, recordId))
+        case _                                                        =>
           Redirect(routes.JourneyRecoveryController.onPageLoad().url)
       }
   }
 
   def onSubmit(mode: Mode, recordId: String): Action[AnyContent] =
     (identify andThen getData andThen requireData).async { implicit request =>
-      val categorisationInfoOpt = request.userAnswers.get(RecordCategorisationsQuery).flatMap(_.records.get(recordId))
-
-      val originalComcodeOpt = categorisationInfoOpt.flatMap(_.originalCommodityCode)
-      val latestComcodeOpt   = categorisationInfoOpt.map(_.commodityCode)
-
-      originalComcodeOpt match {
-        case Some(shortCommodity) if commodityCodeSansTrailingZeros(shortCommodity).length == 6 =>
-          val shortCode = commodityCodeSansTrailingZeros(shortCommodity)
+      val shortComcodeOpt         = getShortCommodityCodeOpt(recordId, request, validLength)
+      val currentlyCategorisedOpt =
+        request.userAnswers
+          .get(RecordCategorisationsQuery)
+          .flatMap(_.records.get(recordId))
+          .map(_.commodityCode.padTo(maxValidLength, "0").mkString)
+
+      shortComcodeOpt match {
+        case Some(shortComcode) if shortComcode.length == validLength =>
           form
             .bindFromRequest()
             .fold(
-              formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode, shortCode, recordId))),
+              formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode, shortComcode, recordId))),
               value => {
-                val longCommodityCode   = s"$shortCode$value"
-                val shouldRedirectToCya = latestComcodeOpt.exists(latest =>
-                  shortCode.concat(value) == latest &&
-                    categorisationInfoOpt.exists(_.latestDoesNotMatchOriginal)
+                val longCommodityCode   = (shortComcode + value).padTo(maxValidLength, "0").mkString
+                val shouldRedirectToCya = currentlyCategorisedOpt.contains(longCommodityCode) && mode == CheckMode
```

Comment:
this feels like it should just be comparing the value with what's already in userAnswers why are you looking at the recordCategorisations query?

Reply:
Because it's possible that a user is in CheckMode, changes their answer, then selects "no" on HasCorrectGoods, comes back and then just submits that answer again, redirecting them to CYA with a different answer that they didn't ever say yes to. The source of truth for commodity codes that have been categorised is recordCategorisations, which is updated with the answer the user puts in if they say "yes". I hope that makes sense?